### ISSUE TITLE
Provide reference date to client on page load

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#6782]
## Feature Description

The client currently references the current date using a `getReferenceDate` selector on the `core/user` store. This was initially introduced to provide a date in tests to make resulting report requests/data consistent. Since then it has been used as a more general reference to the current date and more recently for date-based logic.

Until now, this has been set using the current date on the client (if not already set). In #6549, we decided to make this something that could be externally controllable for use in QA of more upcoming date-based changes.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The [initial `referenceDate` state](https://github.com/google/site-kit-wp/blob/ba1e82b3e7a5087fc61ba9e820878bf37478995a/assets/js/googlesitekit/datastore/user/date-range.js#L38) should be set by an optional server-side value provided on page-load
	* By default, no value should be passed, preserving the same behavior as today
	* If provided, the server value should take precedence over the default value provided by the client

## Implementation Brief

Within `includes/Core/Assets/Assets.php`, in the `get_inline_base_data` method:
- Add a new `referenceDate` field to the inline data array. The value of this new field should be the result of applying a new filter named `googlesitekit_reference_date`, with the argument `null`. This will allow the value to be overridden in future by adding this filter, with it otherwise defaulting to `null`.

Within `assets/js/googlesitekit/datastore/user/date-range.js`, in the `getReferenceDate()` selector:
- If `global._googlesitekitBaseData.referenceDate` is a non-`null` value, return it instead of the `referenceDate` value from state.

### Test Coverage

- Add JS and PHP test coverage for these changes.
- Existing tests should continue to pass.

## QA Brief

* Without any changes, check the output of `getReferenceDate` selector, it should be current date. 
* `googlesitekit.data.select('core/user').getReferenceDate();`
* Now, add the following php snippet to a `mu-plugin` and reload the page. 
````php
<?php
add_filter('googlesitekit_reference_date' , function() {
    return '2023-09-01';
});
````
* The output of `getReferenceDate` should change to `2023-09-01` as set in the snippet. 

## Changelog entry

* Provide reference date to client on page load, in order to be able to model given dates for testing.

**Comment by tofumatt** [comment]
IB ✅

**Comment by kuasha420** [comment]
@techanvil The `googlesitekit_inline_base_data` already has a filter that can be used to add the `referenceDate` to the `base_data` from the tester plugin. Is there any value in adding the `referenceDate` and the specific filter here in the PHP part or just doing the JS part is sufficient?

**Comment by aaemnnosttv** [comment]
@kuasha420 you're correct in that it is not strictly necessary, but I think there is value in adding a filter for this specifically.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- The output of getReferenceDate selector is the current date.

When I have used the code in a mu-plugin:
- The output of getReferenceDate is now 2023-09-01 as set in the snippet.

![image](https://user-images.githubusercontent.com/73545194/229947294-f7349be0-4184-4c83-b911-fe1b3803253a.png)
![image](https://user-images.githubusercontent.com/73545194/229947299-ce1f8614-e403-4c16-958c-b2c217c82689.png)

**Comment by aaemnnosttv** [comment]
## ⚠️ Approval

The AC isn't met here which calls for the initial state to be set by the data passed here, but the override happens in the selector.

This was done according to the IB so it should have been flagged earlier.

Let's update this to be as specified in the AC to keep selectors returning based on internal state only.

cc: @tofumatt

**Comment by aaemnnosttv** [comment]
Fixed in #6829 👍
